modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/aware/SegmentArchiveSizeStorage.java

```diff
@@ -17,65 +17,132 @@
+     */
+    @Nullable private final Map<Long, Long> segmentSize;
 
-    /** Flag of interrupt waiting on this object. */
-    private volatile boolean interrupted;
+    /**
+     * Segment reservations storage.
+     * {@code null} if {@link #maxWalArchiveSize} == {@link DataStorageConfiguration#UNLIMITED_WAL_ARCHIVE}.
+     */
+    @Nullable private final SegmentReservationStorage reservationStorage;
 
     /**
-     * Adding current WAL archive size in bytes.
+     * Constructor.
      *
-     * @param size Size in bytes.
+     * @param minWalArchiveSize Minimum size of the WAL archive in bytes.
+     * @param maxWalArchiveSize Maximum size of the WAL archive in bytes
+     *      or {@link DataStorageConfiguration#UNLIMITED_WAL_ARCHIVE}.
+     * @param reservationStorage Segment reservations storage.
      */
-    synchronized void addCurrentSize(long size) {
-        curr += size;
+    public SegmentArchiveSizeStorage(
+        long minWalArchiveSize,
+        long maxWalArchiveSize,
+        SegmentReservationStorage reservationStorage
+    ) {
+        this.minWalArchiveSize = minWalArchiveSize;
+        this.maxWalArchiveSize = maxWalArchiveSize;
 
-        if (size > 0)
-            notifyAll();
+        if (maxWalArchiveSize != UNLIMITED_WAL_ARCHIVE) {
+            segmentSize = new TreeMap<>();
+            this.reservationStorage = reservationStorage;
+        }
+        else {
+            segmentSize = null;
+            this.reservationStorage = null;
+        }
     }
 
     /**
-     * Adding reserved WAL archive size in bytes.
-     * Defines a hint to determine if the maximum size is exceeded before a new segment is archived.
+     * Adding the WAL segment size in the archive.
      *
-     * @param size Size in bytes.
+     * @param idx Absolut segment index.
+     * @param curr Current WAL archive size in bytes.
      */
-    synchronized void addReservedSize(long size) {
-        reserved += size;
+    void addSize(long idx, long curr) {
+        long releaseIdx = -1;
+
+        synchronized (this) {
+            this.curr += curr;
+
+            if (segmentSize != null) {
+                segmentSize.compute(idx, (i, size) -> {
+                    long res = (size == null ? 0 : size) + curr;
+
+                    return res == 0 ? null : res;
+                });
+            }
+
+            if (curr > 0) {
+                if (segmentSize != null && this.curr >= maxWalArchiveSize) {
+                    long size = 0;
+
+                    for (Map.Entry<Long, Long> e : segmentSize.entrySet()) {
+                        releaseIdx = e.getKey();
+
+                        if (this.curr - (size += e.getValue()) < minWalArchiveSize)
+                            break;
+                    }
+                }
 
-        if (size > 0)
-            notifyAll();
+                notifyAll();
+            }
+        }
+
+        if (releaseIdx != -1) {
+            assert reservationStorage != null;
```

Comment:
Assert doesn't provide information on which `releaseIdx` it has failed. Please add a message to the assert clause.